crates/provider/src/native.rs

```diff
@@ -2,531 +2,920 @@ use std::{
+use uuid::Uuid;
 
-use super::Provider;
 use crate::{
     errors::ProviderError,
-    shared::{
-        constants::{DEFAULT_DATA_DIR, DEFAULT_REMOTE_DIR, LOCALHOST, P2P_PORT},
-        types::{FileMap, NativeRunCommandOptions, Process, RunCommandResponse},
-    },
+    shared::constants::{DEFAULT_TMP_DIR, NODE_CONFIG_DIR, NODE_DATA_DIR},
+    DynNamespace, DynNode, ExecutionResult, Provider, ProviderCapabilities, ProviderNamespace,
+    ProviderNode, RunCommandOptions, RunScriptOptions, SpawnNodeOptions, SpawnTempOptions,
 };
-#[derive(Debug, Clone, PartialEq)]
-pub struct NativeProvider<T: FileSystem + Send + Sync> {
-    // Namespace of the client (isolation directory)
-    namespace: String,
-    // TODO: re-iterate, since we are creating the config with the sdk
-    // Path where configuration relies, all the `files` are accessed relative to this.
-    // config_path: String,
-    // Command to use, e.g "bash"
-    command: String,
-    // Temporary directory, root directory for the network
-    tmp_dir: String,
-    remote_dir: String,
-    data_dir: String,
-    process_map: HashMap<String, Process>,
-    filesystem: T,
-}
-
-impl<T: FileSystem + Send + Sync> NativeProvider<T> {
-    /// Zombienet `native` provider allows to run the nodes as a local process in the local environment
-    /// params:
-    ///   namespace:  Namespace of the client
-    ///   config_path: Path where configuration relies
-    ///   tmp_dir: Temporary directory where files will be placed
-    ///   filesystem: Filesystem to use (std::fs::FileSystem, mock etc.)
-    pub fn new(
-        namespace: impl Into<String>,
-        // config_path: impl Into<String>,
-        tmp_dir: impl Into<String>,
-        filesystem: T,
-    ) -> Self {
-        let tmp_dir = tmp_dir.into();
-        let process_map: HashMap<String, Process> = HashMap::new();
-
-        Self {
-            namespace: namespace.into(),
-            // config_path: config_path.into(),
-            remote_dir: format!("{}{}", &tmp_dir, DEFAULT_REMOTE_DIR),
-            data_dir: format!("{}{}", &tmp_dir, DEFAULT_DATA_DIR),
-            command: "bash".into(),
-            tmp_dir,
-            process_map,
-            filesystem,
-        }
-    }
 
-    fn get_process_by_node_name(&self, node_name: &str) -> Result<&Process, ProviderError> {
-        self.process_map
-            .get(node_name)
-            .ok_or(ProviderError::MissingNodeInfo(
-                node_name.to_owned(),
-                "process".into(),
-            ))
-    }
-}
-
-pub struct Node {}
-
-#[async_trait]
-impl<T> Provider for NativeProvider<T>
+pub struct NativeProviderOptions<FS>
 where
-    T: FileSystem + Send + Sync,
+    FS: FileSystem + Send + Sync,
 {
-    type Node = Node;
+    filesystem: FS,
+    tmp_dir: Option<String>,
+}
 
-    fn require_image() -> bool {
-        false
-    }
+#[derive(Debug)]
+struct NativeProviderInner<FS: FileSystem + Send + Sync + Clone> {
+    capabilities: ProviderCapabilities,
```

Comment:
Do we need to add this fields into the `inner`, I think `capabilities`, `tmp_dir` and `filesystem` could be part of the `NativeProvider` and only put the namespace into the inner struct, since those are mostly initialized in `new` and will not change.